# Homework 03: Explore Titanic Data with `pandas`
Total 14 Questions with 5 out 100 credit points.

Homework 03 is due on 09/18/2021 Saturday 11:59pm CST. Without a reason, later submission within one week will receive a penalty of 20% of total credit scores, and any submission after one week post due will receive no credit.

Notes:  
* The homework will use data sourced from [Kaggle Titanic: Machine Learning from Disaster](http://www.kaggle.com/c/titanic-gettingStarted/data) survivorship prediction task. If you have not used Kaggle before, you need to register an account to pull the data. For your convenience, the data set is provided as an attachement for this HW. Download the nobook file and data file and save in the same path before import the data    

* For data dictionaries, please check the attached "Data Dictionary" file. Read the data dictionary before your homework, as you have to understand what each column represents.
* It is recommended to download the `.ipynb` and data files in the same location so it is convinient to read in the csv files.  
* Answer to some of the questions are not given directly in class and requires self-research effort.

**Variable Dictionary:**
* pclass: A proxy for socio-economic status (SES), 1st = Upper, 2nd = Middle, 3rd = Lower

* age: Age is fractional if less than 1. If the age is estimated, is it in the form of xx.5

* sibsp: The dataset defines family relations in this way...
* Sibling = brother, sister, stepbrother, stepsister
* Spouse = husband, wife (mistresses and fiancés were ignored)

* Parch: The dataset defines family relations in this way...
* Parent = mother, father
* Child = daughter, son, stepdaughter, stepson
* Some children travelled only with a nanny, therefore parch=0 for them.

**Q1** (2 points) Use pandas to import and explore data

In [1]:
import numpy as np

import pandas as pd # your code to import the pandas package with an alias name of "pd"


**Q1-1**.  use `read_csv` function to read in the `titanic_train.csv` as `df`

In [2]:
df = pd.read_csv('./titanic_train.csv') # provide your answer here


**Q1-2**. Verify that you get `891` lines of data (print the number of rows)

In [3]:
len(df.index) # provide your answer here
# or use
# df.shape[0] 

891

**Q1-3**. print the first ten (`head` method) the dataframe `df` 

In [4]:
df.head(n=10) # provide your answer here


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

**Q1-4**.  use `info()` method to get a view of non-missing value and data type

In [17]:
df.info() # provide your answer here


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


**Q1-5**. Use "describe" method to get summary of numeric variables

In [18]:
df.describe() # provide your answer here


PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

**Q1-6**. Uee "count()" or "isnull().sum()" to show how much non-missing values for all the variables. 

In [19]:
print(df.notnull().sum())# provide your answer here


PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64


**Q1-7**. Sampling exercise: create an sample `df_row_sampling` dataframe by indexing/selecting all the even row (row 0,2,4...) from dataframe `df`.  
At the same time drop the `Name` column. For `Name` variable, in the future modeling effort, may not be the best predictor of Survivalship.    

hint: using `iloc` with `step` argument `([... :... : 2])` for row manipulation;  `.drop` method (specifying `axis=1`) for column manipulation

In [20]:
# provide your answer here
df_row_sampling = df.iloc[:: 2]
df_row_sampling.drop('Name', axis=1) 


PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch  \
0              1         0       3    male  22.0      1      0   
2              3         1       3  female  26.0      0      0   
4              5         0       3    male  35.0      0      0   
6              7         0       1    male  54.0      0      0   
8              9         1       3  female  27.0      0      2   
..           ...       ...     ...     ...   ...    ...    ...   
882          883         0       3  female  22.0      0      0   
884          885         0       3    male  25.0      0      0   
886          887         0       2    male  27.0      0      0   
888          889         0       3  female   NaN      1      2   
890          891         0       3    male  32.0      0      0   

               Ticket     Fare Cabin Embarked  
0           A/5 21171   7.2500   NaN        S  
2    STON/O2. 3101282   7.9250   NaN        S  
4              373450   8.0500   NaN        S  
6               17463  51.8625   E46        S  
8              347742  11.1333   NaN        S  
..                ...      ...   ...      ...  
882              7552  10.5167   NaN        S  
884   SOTON/OQ 392076   7.0500   NaN        S  
886            211536  13.0000   NaN        S  
888        W./C. 6607  23.4500   NaN        S  
890            370376   7.7500   NaN        Q  

[446 rows x 11 columns]

**Q1-8**. Create an sample `df_age_sex_sp` dataframe (from orignal `df` dataframe) with filtering records (rows) for `age larger than 50 and all females`
* Hints: select rows with "&" operator of 2 boolean logics on "Age" and "Sex"   

In [37]:
# provide your answer here
# df_age_sex_sp = df[(df.Age > 50) & (df.Sex=='female')]
# df_age_sex_sp 
# OR
df_age_sex_sp = df.query("Age > 50 & Sex == 'female'")
df_age_sex_sp


PassengerId  Survived  Pclass  \
11            12         1       1   
15            16         1       2   
195          196         1       1   
268          269         1       1   
275          276         1       1   
366          367         1       1   
483          484         1       3   
496          497         1       1   
513          514         1       1   
571          572         1       1   
591          592         1       1   
765          766         1       1   
772          773         0       2   
774          775         1       2   
820          821         1       1   
829          830         1       1   
879          880         1       1   

                                                  Name     Sex   Age  SibSp  \
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
195                               Lurette, Miss. Elise  female  58.0      0   
268      Graham, Mrs. William Thompson (Edith Junkins)  female  58.0      0   
275                  Andrews, Miss. Kornelia Theodosia  female  63.0      1   
366   Warren, Mrs. Frank Manley (Anna Sophia Atkinson)  female  60.0      1   
483                             Turkula, Mrs. (Hedwig)  female  63.0      0   
496                     Eustis, Miss. Elizabeth Mussey  female  54.0      1   
513     Rothschild, Mrs. Martin (Elizabeth L. Barrett)  female  54.0      1   
571      Appleton, Mrs. Edward Dale (Charlotte Lamson)  female  53.0      2   
591    Stephenson, Mrs. Walter Bertram (Martha Eustis)  female  52.0      1   
765               Hogeboom, Mrs. John C (Anna Andrews)  female  51.0      1   
772                                  Mack, Mrs. (Mary)  female  57.0      0   
774              Hocking, Mrs. Elizabeth (Eliza Needs)  female  54.0      1   
820  Hays, Mrs. Charles Melville (Clara Jennings Gr...  female  52.0      1   
829          Stone, Mrs. George Nelson (Martha Evelyn)  female  62.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   

     Parch       Ticket      Fare Cabin Embarked  Tol_Family_No Age_group  
11       0       113783   26.5500  C103        S              0  (40, 60]  
15       0       248706   16.0000   NaN        S              0  (40, 60]  
195      0     PC 17569  146.5208   B80        C              0  (40, 60]  
268      1     PC 17582  153.4625  C125        S              1  (40, 60]  
275      0        13502   77.9583    D7        S              1  (60, 80]  
366      0       110813   75.2500   D37        C              1  (40, 60]  
483      0         4134    9.5875   NaN        S              0  (60, 80]  
496      0        36947   78.2667   D20        C              1  (40, 60]  
513      0     PC 17603   59.4000   NaN        C              1  (40, 60]  
571      0        11769   51.4792  C101        S              2  (40, 60]  
591      0        36947   78.2667   D20        C              1  (40, 60]  
765      0        13502   77.9583   D11        S              1  (40, 60]  
772      0  S.O./P.P. 3   10.5000   E77        S              0  (40, 60]  
774      3        29105   23.0000   NaN        S              4  (40, 60]  
820      1        12749   93.5000   B69        S              2  (40, 60]  
829      0       113572   80.0000   B28      NaN              0  (60, 80]  
879      1        11767   83.1583   C50        C              1  (40, 60]

**Q1-9**. In the `df` Dataframe, create a new column `Tol_Family_No` to calculate the total family number: by adding numbers in sibling (`SibSp`) and parent (`Parch`) columns

In [22]:
df["Tol_Family_No"] = df["Parch"] + df["SibSp"] # provide your answer here


In [23]:
df.head(n=5)



PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Tol_Family_No  
0      0         A/5 21171   7.2500   NaN        S              1  
1      0          PC 17599  71.2833   C85        C              1  
2      0  STON/O2. 3101282   7.9250   NaN        S              0  
3      0            113803  53.1000  C123        S              1  
4      0            373450   8.0500   NaN        S              0

**Q1-10** The following statement tells the decomsition of the people survied, a.k.a column `Survived` as of 0 (not survived) or 1 (survived). Create a similar statement using `groupby` function to further breakdown by `Sex`. In other words, add additional `sex` dimention. The desired output should be similar to the following:   

`Sex    Survided  
female   0 	      81 
         1          233 
male     0 	      468
         1          109       `  
`Name: Survived, dtype: int64 `


In [11]:
#Example code, Attention : "groupby" on column only take either 1 single column or a list of columns 


Survived
0    549
1    342
Name: Survived, dtype: int64

In [24]:
df.groupby(['Sex', 'Survived'])["Survived"].size()# provide your answer here


Sex     Survived
female  0            81
        1           233
male    0           468
        1           109
Name: Survived, dtype: int64

**Q2** (1 point). Bin Continuous Values (`Age`) in `df` into Discrete Intervals (create a new column in `df` called `Age_group`) : `["20-", "20-40", "40-60", "60-80","80+" ]`  
*hint: you can choose to use `pd.cut(column, bins)` method*


In [25]:
#With defined age group levels of Age_level as below
Age_bin = [i for i in range(int(min(df.Age)), 100, 20)]
print(Age_bin)

# you are not required to use levels defined in 'Age_level` create a new column as of "Age_grp".

# provide your answer here
df['Age_group'] = pd.cut(df['Age'], bins=[0, 20, 40, 60, 80])
df.Age_group.unique()



[0, 20, 40, 60, 80]


[(20.0, 40.0], NaN, (40.0, 60.0], (0.0, 20.0], (60.0, 80.0]]
Categories (4, interval[int64]): [(0, 20] < (20, 40] < (40, 60] < (60, 80]]

In [26]:
print(df.shape)
df.head()


(891, 14)


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Tol_Family_No Age_group  
0      0         A/5 21171   7.2500   NaN        S              1  (20, 40]  
1      0          PC 17599  71.2833   C85        C              1  (20, 40]  
2      0  STON/O2. 3101282   7.9250   NaN        S              0  (20, 40]  
3      0            113803  53.1000  C123        S              1  (20, 40]  
4      0            373450   8.0500   NaN        S              0  (20, 40]

**Q3** (1 point). Using `df`, create a pivot table to count `survived`=1 by `Age_grp` and `Sex`. The desired output is similar to the following:

`Sex       female  male
Age_grp               
(0, 20]       53    29
(20, 40]     107    46
(40, 60]      34    16
(60, 80]       3     2    `

In [41]:
# provide your answer here)

Survived_by_AgeSex = pd.pivot_table(data=df,index=['Age_group'], columns=['Sex'], aggfunc={'Survived':np.sum})
print(Survived_by_AgeSex)
    

          Survived     
Sex         female male
Age_group              
(0, 20]         53   29
(20, 40]       107   46
(40, 60]        34   16
(60, 80]         3    2


**Q4** (1 point). The aim of this task is to stack 'titanic_train.cvs' and 'titanic_test.cvs' into 1 table and reset index using `PassengerId`.   

**Q4-1** first read in `titanic_test` as df1

In [38]:
df1 = pd.read_csv('titanic_test.csv') # provide your answer here


In [39]:
df1.head(2)


PassengerId  Pclass                              Name     Sex   Age  SibSp  \
0          892       3                  Kelly, Mr. James    male  34.5      0   
1          893       3  Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   

   Parch  Ticket    Fare Cabin Embarked  
0      0  330911  7.8292   NaN        Q  
1      0  363272  7.0000   NaN        S

**Q4-2** First, list the column names of both train and test dataset, seperately

In [40]:
#to print columns of df and df1, provide your answer here
print(df.columns)
df1.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'Tol_Family_No',
       'Age_group'],
      dtype='object')


Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

Note the only differnce is the column named `Survived` which only exist in train data but not the test data. This is because the test data is used to predict the `Survived` field using machine learning modeling techniques (which we will cover in later class). Sometimes we will want to join the tables to take a look at some characteristics of the full sample to investigation purpose. This is why we may want to stack the data set.

**Q4-3** First, let's create a new column (use `insert` method) `Survived` of -1 for missing values into test dataset as place holder.   
Make sure it is inserted as the **second** column with column index of 1 
Hint: use `insert` method modifies the original data `df1`

In [42]:
# provide your answer here
df1.insert(1, "Survived", "-1") 

df1

PassengerId Survived  Pclass  \
0            892       -1       3   
1            893       -1       3   
2            894       -1       2   
3            895       -1       3   
4            896       -1       3   
..           ...      ...     ...   
413         1305       -1       3   
414         1306       -1       1   
415         1307       -1       3   
416         1308       -1       3   
417         1309       -1       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                                Kelly, Mr. James    male  34.5      0      0   
1                Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                       Myles, Mr. Thomas Francis    male  62.0      0      0   
3                                Wirz, Mr. Albert    male  27.0      0      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   
..                                            ...     ...   ...    ...    ...   
413                            Spector, Mr. Woolf    male   NaN      0      0   
414                  Oliva y Ocana, Dona. Fermina  female  39.0      0      0   
415                  Saether, Mr. Simon Sivertsen    male  38.5      0      0   
416                           Ware, Mr. Frederick    male   NaN      0      0   
417                      Peter, Master. Michael J    male   NaN      1      1   

                 Ticket      Fare Cabin Embarked  
0                330911    7.8292   NaN        Q  
1                363272    7.0000   NaN        S  
2                240276    9.6875   NaN        Q  
3                315154    8.6625   NaN        S  
4               3101298   12.2875   NaN        S  
..                  ...       ...   ...      ...  
413           A.5. 3236    8.0500   NaN        S  
414            PC 17758  108.9000  C105        C  
415  SOTON/O.Q. 3101262    7.2500   NaN        S  
416              359309    8.0500   NaN        S  
417                2668   22.3583   NaN        C  

[418 rows x 12 columns]

In [43]:
#take a look again at the `df1` with the additional column
print(df1.shape)
df1.head()


(418, 12)


PassengerId Survived  Pclass                                          Name  \
0          892       -1       3                              Kelly, Mr. James   
1          893       -1       3              Wilkes, Mrs. James (Ellen Needs)   
2          894       -1       2                     Myles, Mr. Thomas Francis   
3          895       -1       3                              Wirz, Mr. Albert   
4          896       -1       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   

      Sex   Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0    male  34.5      0      0   330911   7.8292   NaN        Q  
1  female  47.0      1      0   363272   7.0000   NaN        S  
2    male  62.0      0      0   240276   9.6875   NaN        Q  
3    male  27.0      0      0   315154   8.6625   NaN        S  
4  female  22.0      1      1  3101298  12.2875   NaN        S

**Q4-4** Explore using `concat`,`merge`, `append`, or `melt` function to stack the two data set vertically. Try different settings and see which function(s) works the best for you. 
Name the final results dataset `df_full`. `df_full` should have `len(df)` + `len(df1)` = 1309 rows and `len(df_full.columns)` = 14 columns.

In [29]:
# provide your answer here
df_full = pd.concat([df,df1])
df_full

PassengerId Survived  Pclass  \
0              1        0       3   
1              2        1       1   
2              3        1       3   
3              4        1       1   
4              5        0       3   
..           ...      ...     ...   
413         1305       -1       3   
414         1306       -1       1   
415         1307       -1       3   
416         1308       -1       3   
417         1309       -1       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
413                                 Spector, Mr. Woolf    male   NaN      0   
414                       Oliva y Ocana, Dona. Fermina  female  39.0      0   
415                       Saether, Mr. Simon Sivertsen    male  38.5      0   
416                                Ware, Mr. Frederick    male   NaN      0   
417                           Peter, Master. Michael J    male   NaN      1   

     Parch              Ticket      Fare Cabin Embarked  Tol_Family_No  \
0        0           A/5 21171    7.2500   NaN        S            1.0   
1        0            PC 17599   71.2833   C85        C            1.0   
2        0    STON/O2. 3101282    7.9250   NaN        S            0.0   
3        0              113803   53.1000  C123        S            1.0   
4        0              373450    8.0500   NaN        S            0.0   
..     ...                 ...       ...   ...      ...            ...   
413      0           A.5. 3236    8.0500   NaN        S            NaN   
414      0            PC 17758  108.9000  C105        C            NaN   
415      0  SOTON/O.Q. 3101262    7.2500   NaN        S            NaN   
416      0              359309    8.0500   NaN        S            NaN   
417      1                2668   22.3583   NaN        C            NaN   

    Age_group  
0    (20, 40]  
1    (20, 40]  
2    (20, 40]  
3    (20, 40]  
4    (20, 40]  
..        ...  
413       NaN  
414       NaN  
415       NaN  
416       NaN  
417       NaN  

[1309 rows x 14 columns]

In [30]:
#check your output df
print(df_full.shape)
df_full.tail()

(1309, 14)


PassengerId Survived  Pclass                          Name     Sex   Age  \
413         1305       -1       3            Spector, Mr. Woolf    male   NaN   
414         1306       -1       1  Oliva y Ocana, Dona. Fermina  female  39.0   
415         1307       -1       3  Saether, Mr. Simon Sivertsen    male  38.5   
416         1308       -1       3           Ware, Mr. Frederick    male   NaN   
417         1309       -1       3      Peter, Master. Michael J    male   NaN   

     SibSp  Parch              Ticket      Fare Cabin Embarked  Tol_Family_No  \
413      0      0           A.5. 3236    8.0500   NaN        S            NaN   
414      0      0            PC 17758  108.9000  C105        C            NaN   
415      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S            NaN   
416      0      0              359309    8.0500   NaN        S            NaN   
417      1      1                2668   22.3583   NaN        C            NaN   

    Age_group  
413       NaN  
414       NaN  
415       NaN  
416       NaN  
417       NaN

**Q4-5** last, in the newly created `df_full` data, reset the index using `PassengerId` column using `.set_index()` or other method you favor. Remeber to drop the ``PassengerId`` to end up with df_full with 13 columns. 

In [35]:
# provide your answer here
df_full=df_full.set_index('PassengerId')

In [36]:
#check your output df
print(df_full.shape)
df_full.head()

(1309, 13)


Survived  Pclass  \
PassengerId                    
1                  0       3   
2                  1       1   
3                  1       3   
4                  1       1   
5                  0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500   NaN        S   
2                1      0          PC 17599  71.2833   C85        C   
3                0      0  STON/O2. 3101282   7.9250   NaN        S   
4                1      0            113803  53.1000  C123        S   
5                0      0            373450   8.0500   NaN        S   

             Tol_Family_No Age_group  
PassengerId                           
1                      1.0  (20, 40]  
2                      1.0  (20, 40]  
3                      0.0  (20, 40]  
4                      1.0  (20, 40]  
5                      0.0  (20, 40]